src/macos/monitor.rs

```diff
@@ -2,115 +2,175 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
-use std::io;
-use std::sync::mpsc::{channel, Sender, Receiver, TryIter};
-use std::thread;
-
-use super::iohid::*;
-use super::iokit::*;
-use core_foundation_sys::runloop::*;
-use runloop::RunLoop;
-
 extern crate log;
 extern crate libc;
-use libc::c_void;
 
-pub enum Event {
-    Add(IOHIDDeviceRef),
-    Remove(IOHIDDeviceRef),
+use core_foundation_sys::base::*;
+use core_foundation_sys::runloop::*;
+use libc::c_void;
+use platform::iohid::*;
+use platform::iokit::*;
+use runloop::RunLoop;
+use std::{io, slice};
+use std::collections::HashMap;
+use std::sync::mpsc::{channel, Receiver, Sender};
+use util::io_err;
+
+struct DeviceData {
+    tx: Sender<Vec<u8>>,
+    runloop: RunLoop,
 }
 
-pub struct Monitor {
-    // Receive events from the thread.
-    rx: Receiver<Event>,
-    // Handle to the thread loop.
-    thread: RunLoop,
+pub struct Monitor<F>
+where
+    F: Fn(IOHIDDeviceRef, Receiver<Vec<u8>>, &Fn() -> bool) + Sync,
+{
+    manager: IOHIDManagerRef,
+    // Keep alive until the monitor goes away.
+    _matcher: IOHIDDeviceMatcher,
+    map: HashMap<IOHIDDeviceRef, DeviceData>,
+    new_device_cb: F,
 }
 
-impl Monitor {
-    pub fn new() -> io::Result<Self> {
-        let (tx, rx) = channel();
+impl<F> Monitor<F>
+where
+    F: Fn(IOHIDDeviceRef, Receiver<Vec<u8>>, &Fn() -> bool) + Sync + 'static,
+{
+    pub fn new(new_device_cb: F) -> Self {
+        let manager = unsafe { IOHIDManagerCreate(kCFAllocatorDefault, kIOHIDManagerOptionNone) };
+
+        // Match FIDO devices only.
+        let _matcher = IOHIDDeviceMatcher::new();
```

Comment:
btw, I'm super impressed that you figured out how to make this work. I fought it ... too long.

Reply:
It took... at lot of time :)